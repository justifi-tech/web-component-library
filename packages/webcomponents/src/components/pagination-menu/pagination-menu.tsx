import { Component, h, Host, Prop } from '@stencil/core';
import { ExtendedPagingDefaults, ExtendedPagingInfo } from '../../api/Pagination';

export interface PaginationProps {
  paging: ExtendedPagingInfo;
  params?: {};
}

@Component({
  tag: 'pagination-menu'
})

export class PaginationMenu {
  @Prop() paging: PaginationProps['paging'] = ExtendedPagingDefaults;
  @Prop() params: PaginationProps['params'] = {};

  render() {
    return (
      <Host>
        <nav aria-label="Table pagination" class="d-flex justify-content-end gap-3">
          <ul class="pagination m-0" part="pagination">
            <li class={`page-item ${this.paging.has_previous ? '' : ' disabled'}`} part="page-item">
              <a href="#"
                class="page-link"
                onClick={(e) => {
                  e.preventDefault();
                  this.paging.handleClickPrevious(this.paging.start_cursor);
                }}
                part={`page-link${this.paging.has_previous ? '' : ' page-link-disabled'}`}>
                <span class="me-1" part="page-link-arrow">&laquo;</span>
                <span part="page-button-text">Previous</span>
              </a>
            </li>
            <li class={`page-item ${this.paging.has_next ? '' : ' disabled'}`}>
              <a href="#"
                class="page-link"
                onClick={(e) => {
                  e.preventDefault();
                  this.paging.handleClickNext(this.paging.end_cursor);
                }}
                part={`page-link${this.paging.has_next ? '' : ' page-link-disabled'}`}>
                <span part="page-link-text">Next</span>
                <span class="ms-1" part="page-link-arrow">&raquo;</span>
              </a>
            </li>
          </ul>
        </nav>
      </Host>
    );
  }
}
