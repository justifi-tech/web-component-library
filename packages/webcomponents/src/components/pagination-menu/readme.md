# pagination-menu



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type                 | Default                  |
| -------- | --------- | ----------- | -------------------- | ------------------------ |
| `paging` | --        |             | `ExtendedPagingInfo` | `ExtendedPagingDefaults` |
| `params` | --        |             | `{}`                 | `{}`                     |


## Shadow Parts

| Part                 | Description |
| -------------------- | ----------- |
| `"page-arrow"`       |             |
| `"page-button-text"` |             |


## Dependencies

### Used by

 - [payments-list-core](../payments-list)
 - [payouts-list-core](../payouts-list)
 - [terminals-list-core](../terminals-list)

### Graph
```mermaid
graph TD;
  payments-list-core --> pagination-menu
  payouts-list-core --> pagination-menu
  terminals-list-core --> pagination-menu
  style pagination-menu fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
