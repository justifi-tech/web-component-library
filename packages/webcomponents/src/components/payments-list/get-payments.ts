import { Payment } from '../../api';

export const makeGetPayments =
  ({ id, authToken, service }) =>
  async ({ params, onSuccess, onError }) => {
    try {
      const response = await service.fetchPayments(id, authToken, params);

      if (!response.error) {
        const pagingInfo = {
          ...response.page_info,
        };

        const payments =
          response.data?.map((dataItem) => new Payment(dataItem)) || [];

        onSuccess({ payments, pagingInfo });
      } else {
        const responseError =
          typeof response.error === 'string'
            ? response.error
            : response.error.message;

        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };
