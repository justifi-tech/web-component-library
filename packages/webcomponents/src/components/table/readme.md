# justifi-table



<!-- Auto Generated Below -->


## Properties

| Property                  | Attribute       | Description | Type                     | Default                  |
| ------------------------- | --------------- | ----------- | ------------------------ | ------------------------ |
| `columnData` _(required)_ | --              |             | `(string \| string[])[]` | `undefined`              |
| `entityId`                | --              |             | `string[]`               | `undefined`              |
| `errorMessage`            | `error-message` |             | `string`                 | `''`                     |
| `loading`                 | `loading`       |             | `boolean`                | `true`                   |
| `paging`                  | --              |             | `ExtendedPagingInfo`     | `ExtendedPagingDefaults` |
| `params`                  | `params`        |             | `any`                    | `{}`                     |
| `rowClickHandler`         | --              |             | `(e: any) => any`        | `undefined`              |
| `rowData`                 | --              |             | `any[]`                  | `[]`                     |


## Shadow Parts

| Part                | Description |
| ------------------- | ----------- |
| `"pagination-bar"`  |             |
| `"table-body"`      |             |
| `"table-cell"`      |             |
| `"table-head"`      |             |
| `"table-head-cell"` |             |
| `"table-head-row"`  |             |


## Dependencies

### Used by

 - [justifi-business-list](../business-list)
 - [justifi-payment-balance-transactions](../payment-balance-transactions)
 - [justifi-proceeds-list](../proceeds-list)
 - [justifi-subaccounts-list](../subaccounts-list)
 - [payments-list-core](../payments-list)
 - [payouts-list-core](../payouts-list)

### Depends on

- [pagination-menu](../pagination-menu)

### Graph
```mermaid
graph TD;
  justifi-table --> pagination-menu
  justifi-business-list --> justifi-table
  justifi-payment-balance-transactions --> justifi-table
  justifi-proceeds-list --> justifi-table
  justifi-subaccounts-list --> justifi-table
  payments-list-core --> justifi-table
  payouts-list-core --> justifi-table
  style justifi-table fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
