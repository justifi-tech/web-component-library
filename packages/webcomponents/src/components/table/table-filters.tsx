import { Component, h, Prop, Event, EventEmitter, Listen } from '@stencil/core';
import { Button } from '../../ui-components';

@Component({
  tag: 'table-filters'
})
export class TableFilters {
  @Prop() params: any;
  @Event() clearParams: EventEmitter<any>;
  @Event() setParamsOnChange: EventEmitter<any>;

  get hasFilters() {
    const paginationKeys = ['after_cursor', 'before_cursor'];
    const paramKeys = Object.keys(this.params);
    const filterKeys = paramKeys.filter((paramKey) => !paginationKeys.includes(paramKey));
    return !!filterKeys.length;
  }

  private clearFilters() {
    this.clearParams.emit({});
  }

  @Listen('setParamsOnChange')
  setParams(event: CustomEvent) {
    this.setParamsOnChange.emit(event.detail);
  }



  render() {
    return (
      <div class="container mt-2">
        <div class="row gy-3 mb-4">
          <slot></slot>
        </div>
        {this.hasFilters && (
          <div class="row">
            <div class="col">
              <Button variant='primary' onClick={() => this.clearFilters()}>
                Clear Filters
              </Button>
            </div>
          </div>
        )}
      </div>
    );
  }
}
