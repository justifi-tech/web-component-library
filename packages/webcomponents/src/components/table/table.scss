@import "reboot";
@import "tables";
@import "buttons";
@import "badge";
@import "grid";
@import "bootstrap-utilities";
@import "pagination";

:host {
  @include host-base-styles;
}
