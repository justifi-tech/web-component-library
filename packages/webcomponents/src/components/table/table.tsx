import { Component, Host, h, Prop } from '@stencil/core';
import { ExtendedPagingDefaults, ExtendedPagingInfo } from '../../api/Pagination';

export interface TableProps {
  loading: boolean;
  errorMessage: string;
  columnData: (string | string[])[];
  rowData: any[];
  paging: ExtendedPagingInfo;
  params: any
}

@Component({
  tag: 'justifi-table',
  styleUrl: './table.scss',
  shadow: true,
})

export class Table {
  @Prop() loading: TableProps['loading'] = true;
  @Prop() errorMessage: TableProps['errorMessage'] = '';
  @Prop() rowData: TableProps['rowData'] = [];
  @Prop() columnData!: TableProps['columnData'];
  @Prop() paging: TableProps['paging'] = ExtendedPagingDefaults;
  @Prop() params: TableProps['params'] = {};
  @Prop() entityId: string[];
  @Prop() rowClickHandler: (e: any) => any;

  showEmptyState() {
    return this.rowData ? this.rowData.length < 1 : true;
  }

  emptyState = () => (
    <tr>
      <td class="empty-state" part="empty-state" colSpan={this.columnData?.length} style={{ textAlign: 'center' }}>No results</td>
    </tr>
  );

  errorState = () => (
    <tr>
      <td class="error-state" part="error-state" colSpan={this.columnData?.length} style={{ textAlign: 'center' }}>
        An unexpected error occurred: {this.errorMessage}
      </td>
    </tr>
  );

  loadingState = () => (
    <tr>
      <td class="loading-state" part="loading-state-cell" colSpan={this.columnData?.length} style={{ textAlign: 'center' }}>
        <div part="loading-state-spinner" class="spinner-border" role="status">
          <span class="visually-hidden">Loading...</span>
        </div>
      </td>
    </tr>
  );

  render() {
    return (
      <Host exportparts="
        table-head,table-head-row,table-head-cell,table-body,table-row,table-row-even,
        table-row-odd,table-cell,loading-state-cell,loading-state-spinner,error-state,
        empty-state,pagination-bar,arrow,arrow-left,arrow-right,button-disabled,
        previous-button-text,next-button-text
      ">
        {this.columnData ?
          <table class="table table-hover">
            <thead class="table-head sticky-top" part="table-head">
              <tr class="table-light" part='table-head-row'>
                {
                  this.columnData?.map((column) =>
                    <th part="table-head-cell" scope="col" title={Array.isArray(column) ? column[1] : ''}>
                      {!Array.isArray(column) ? column : column[0]}
                    </th>
                  )
                }
              </tr>
            </thead>
            <tbody class="table-body" part='table-body'>
              {
                this.loading ? this.loadingState() :
                  this.errorMessage ? this.errorState() :
                    this.showEmptyState() ? this.emptyState() :
                      this.rowData?.map((data, index) => {
                        return (
                          <tr
                            data-row-entity-id={this.entityId[index]}
                            onClick={e => this.rowClickHandler ? this.rowClickHandler(e) : null}
                            part={`table-row${index % 2 ? ' table-row-even' : ' table-row-odd'}`}
                          >
                            {
                              data.map((dataEntry: any) => {
                                return (
                                  !dataEntry?.type
                                    ? <td part="table-cell">{dataEntry}</td>
                                    : <th scope="row" part="table-cell">
                                      <td part="table-cell" innerHTML={dataEntry.value}></td>
                                    </th>
                                )
                              }
                              )
                            }
                          </tr>
                        )
                      })
              }
            </tbody>
            {this.paging &&
              <tfoot class="sticky-bottom">
                <tr class="table-light align-middle">
                  <td part="pagination-bar" colSpan={this.columnData?.length}>
                    <pagination-menu paging={this.paging} params={this.params} />
                  </td>
                </tr>
              </tfoot>
            }
          </table>
          :
          <div data-test-id="empty-error-state">Column data is required</div>}
      </Host>
    )
  }
}
