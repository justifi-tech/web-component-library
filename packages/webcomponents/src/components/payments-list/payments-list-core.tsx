import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { PagingInfo, Payment, PaymentsParams, pagingDefaults } from '../../api';
import { MapPaymentStatusToBadge, formatCurrency, formatDate, formatTime } from '../../utils/utils';
import { ComponentError } from '../../api/ComponentError';
import { tableExportedParts } from '../table/exported-parts';
import { StyledHost, TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import { onFilterChange } from '../../ui-components/filters/utils';

@Component({
  tag: 'payments-list-core'
})
export class PaymentsListCore {
  @State() payments: Payment[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() params: PaymentsParams = {};
  
  @Prop() getPayments: Function;

  @Watch('params')
  @Watch('getPayments')
  updateOnPropChange() {
    this.fetchData();
  }

  @Event({
    eventName: 'payment-row-clicked',
    bubbles: true,
  }) rowClicked: EventEmitter<Payment>;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    if (this.getPayments) {
      this.fetchData();
    }
  }

  fetchData(): void {
    this.loading = true;

    this.getPayments({
      params: this.params,
      onSuccess: ({ payments, pagingInfo }) => {
        this.payments = payments;
        this.paging = pagingInfo;
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    const newParams: any = { ...this.params };
    delete newParams.after_cursor;
    this.params = { ...newParams, before_cursor: beforeCursor };
  };

  handleClickNext = (afterCursor: string) => {
    const newParams: any = { ...this.params };
    delete newParams.before_cursor;
    this.params = { ...newParams, after_cursor: afterCursor };
  };

  rowClickHandler = (e) => {
    const clickedPaymentID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedPaymentID) return;
    this.rowClicked.emit(this.payments.find((payment) => payment.id === clickedPaymentID));
  };

  setParamsOnChange = (name: string, value: string) => {
    let newParams = { [name]: value };
    this.params = onFilterChange(newParams, this.params);
  }

  clearParams = () => {
    this.params = {};
  }

  get entityId() {
    return this.payments.map((payment) => payment.id);
  }

  get columnData() {
    return [
      ['Date', 'The date and time each payment was made'],
      ['Amount', 'The dollar amount of each payment'],
      ['Status', 'The current status of each payment'],
      ['Type', 'The type of each payment'],
      ['Description', 'The payment description, if you provided one'],
      ['Account Holder', 'The name associated with the payment method'],
      ['Payment Method', 'The brand and last 4 digits of the payment method'],
    ];
  }

  get rowData() {
    return this.payments.map((payment) => [
      {
        type: 'head',
        value: `
          <div class="fw-bold">${formatDate(payment.created_at)}</div>
          <div class="fw-bold">${formatTime(payment.created_at)}</div>
        `,
      },
      formatCurrency(payment.amount),
      {
        type: 'inner',
        value: MapPaymentStatusToBadge(payment.status),
      },
      'payment type!',
      payment.description,
      payment.payment_method.payersName,
      payment.payment_method.lastFourDigits,
    ]);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState;
  }

  render() {
    return (
      <StyledHost exportparts={tableExportedParts}>
        <payments-list-filters 
          params={this.params} 
          setParamsOnChange={this.setParamsOnChange}
          clearParams={this.clearParams}
        />
        <div class="table-wrapper">
          <table class="table table-hover">
            <thead class="table-head sticky-top" part="table-head">
              <tr class="table-light text-nowrap" part="table-head-row">
                {this.columnData?.map((column) => (
                  <th part="table-head-cell" scope="col" title={Array.isArray(column) ? column[1] : ""}>
                    {!Array.isArray(column) ? column : column[0]}
                  </th>
                ))}
              </tr>
            </thead>
            <tbody class="table-body" part="table-body">
              <TableLoadingState
                columnSpan={this.columnData.length}
                isLoading={this.loading}
              />
              <TableEmptyState
                isEmpty={this.showEmptyState}
                columnSpan={this.columnData.length}
              />
              <TableErrorState
                columnSpan={this.columnData.length}
                errorMessage={this.errorMessage}
              />
              {this.showRowData &&
                this.rowData.map((data, index) => (
                  <tr
                    data-test-id="table-row"
                    data-row-entity-id={this.entityId[index]}
                    onClick={this.rowClickHandler}
                    part={`table-row ${index % 2 ? "table-row-even" : "table-row-odd"}`}
                  >
                    {data.map((dataEntry: any) => {
                      let nestedHtml = dataEntry?.type;
                      if (nestedHtml) {
                        return <td part="table-cell" innerHTML={dataEntry.value}></td>;
                      } else {
                        return <td part="table-cell">{dataEntry}</td>;
                      }
                    })}
                  </tr>
                ))}
            </tbody>
            {this.paging && (
              <tfoot class="sticky-bottom">
                <tr class="table-light align-middle">
                  <td part="pagination-bar" colSpan={this.columnData?.length}>
                    <pagination-menu
                      paging={{
                        ...this.paging,
                        handleClickPrevious: this.handleClickPrevious,
                        handleClickNext: this.handleClickNext,
                      }}
                      params={this.params}
                    />
                  </td>
                </tr>
              </tfoot>
            )}
          </table>
        </div>
      </StyledHost>
    );
  }
}
