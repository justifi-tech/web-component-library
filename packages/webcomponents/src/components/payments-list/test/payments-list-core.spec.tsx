import { newSpecPage } from '@stencil/core/testing';
import { PaymentsListCore } from '../payments-list-core';
import { Table } from '../../table/table';
import { PaginationMenu } from '../../pagination-menu/pagination-menu';
import mockSuccessResponse from '../../../api/mockData/mockPaymentsSuccess.json';
import { IApiResponseCollection, IPayment } from '../../../api';
import { makeGetPayments } from '../get-payments';

const mockPaymentsResponse = mockSuccessResponse as IApiResponseCollection<IPayment[]>;

describe('payments-list-core', () => {
  it('renders properly with fetched data', async () => {
    const mockPaymentsService = {
      fetchPayments: jest.fn().mockResolvedValue(mockPaymentsResponse),
    };

    const page = await newSpecPage({
      components: [PaymentsListCore, Table, PaginationMenu],
      html: `<payments-list-core></payments-list-core>`,
    });

    page.rootInstance.componentWillLoad = () => { };

    page.rootInstance.getPayments = makeGetPayments({
      id: '123',
      authToken: '123',
      service: mockPaymentsService
    });

    page.rootInstance.fetchData();
    await page.waitForChanges();

    const justifiTable = page.root.querySelector('justifi-table');

    expect(page.rootInstance.payments[0]).toEqual(expect.objectContaining({ account_id: mockPaymentsResponse.data[0].account_id }));
    const rows = justifiTable.shadowRoot.querySelectorAll('[data-test-id="table-row"]');
    expect(rows.length).toBe(mockPaymentsResponse.data.length);
    expect(mockPaymentsService.fetchPayments).toHaveBeenCalled();
    expect(page.root).toMatchSnapshot();
  });

  it('displays an error state on failed data fetch', async () => {
    const mockService = {
      fetchPayments: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getPayments = makeGetPayments({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockService
    });

    const page = await newSpecPage({
      components: [PaymentsListCore, Table, PaginationMenu],
      html: '<payments-list-core></payments-list-core>',
    });

    page.rootInstance.componentWillLoad = () => { };

    page.rootInstance.getPayments = getPayments;

    await page.rootInstance.fetchData();
    await page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Fetch error');
    expect(page.root).toMatchSnapshot();
  });

  it('emits payment-row-clicked event on row click', async () => {
    const mockPaymentsService = {
      fetchPayments: jest.fn().mockResolvedValue(mockPaymentsResponse),
    };

    const page = await newSpecPage({
      components: [PaymentsListCore, Table, PaginationMenu],
      html: `<payments-list-core></payments-list-core>`,
    });

    page.rootInstance.componentWillLoad = () => { };

    page.rootInstance.getPayments = makeGetPayments({
      id: '123',
      authToken: '123',
      service: mockPaymentsService
    });

    page.rootInstance.fetchData();

    await page.waitForChanges();

    const justifiTable = page.root.querySelector('justifi-table');
    const firstRow = justifiTable.shadowRoot.querySelector('[data-test-id="table-row"]') as HTMLElement;
    expect(firstRow).not.toBeNull();

    const spyEvent = jest.fn();
    page.win.addEventListener('payment-row-clicked', spyEvent);

    firstRow.click();
    expect(spyEvent).toHaveBeenCalled();
  });

  it('updates params and refetches data on pagination interaction', async () => {
    const mockPaymentsService = {
      fetchPayments: jest.fn().mockResolvedValue(mockPaymentsResponse),
    };

    const page = await newSpecPage({
      components: [PaymentsListCore, Table, PaginationMenu],
      html: `<payments-list-core></payments-list-core>`,
    });

    page.rootInstance.componentWillLoad = () => { };
    page.rootInstance.getPayments = makeGetPayments({
      id: '123',
      authToken: '123',
      service: mockPaymentsService
    });

    page.rootInstance.fetchData();

    await page.waitForChanges();

    // Assuming handleClickNext is accessible and modifies `params` which triggers a re-fetch
    page.rootInstance.handleClickNext('nextCursor');
    await page.waitForChanges();

    // The mock function should be called 3 times: once for the initial load, twice for when mockGetPayments is set and later after the pagination interaction
    expect(mockPaymentsService.fetchPayments).toHaveBeenCalledTimes(3);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams.after_cursor).toBe('nextCursor');
  });
});
