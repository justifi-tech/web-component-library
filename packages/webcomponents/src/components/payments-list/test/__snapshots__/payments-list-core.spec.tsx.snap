// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payments-list-core displays an error state on failed data fetch 1`] = `
<payments-list-core>
  <div>
    <payments-list-filters></payments-list-filters>
    <div class="table-wrapper">
      <table class="table table-hover">
        <thead class="sticky-top table-head" part="table-head">
          <tr class="table-light text-nowrap" part="table-head-row">
            <th part="table-head-cell" scope="col" title="The date and time each payment was made">
              Date
            </th>
            <th part="table-head-cell" scope="col" title="The dollar amount of each payment">
              Amount
            </th>
            <th part="table-head-cell" scope="col" title="The current status of each payment">
              Status
            </th>
            <th part="table-head-cell" scope="col" title="The type of each payment">
              Type
            </th>
            <th part="table-head-cell" scope="col" title="The payment description, if you provided one">
              Description
            </th>
            <th part="table-head-cell" scope="col" title="The name associated with the payment method">
              Account Holder
            </th>
            <th part="table-head-cell" scope="col" title="The brand and last 4 digits of the payment method">
              Payment Method
            </th>
          </tr>
        </thead>
        <tbody class="table-body" part="table-body">
          <tr>
            <td class="error-state" colspan="7" data-test-id="table-error-state" part="error-state" style="text-align: center;">
              An unexpected error occurred: Fetch error
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="7" part="pagination-bar">
              <pagination-menu>
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination" part="pagination">
                    <li class="disabled page-item" part="page-item">
                      <a class="page-link" href="#" part="page-link page-link-disabled">
                        <span class="me-1" part="page-link-arrow">
                          «
                        </span>
                        <span part="page-button-text">
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-link page-link-disabled">
                        <span part="page-link-text">
                          Next
                        </span>
                        <span class="ms-1" part="page-link-arrow">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</payments-list-core>
`;

exports[`payments-list-core renders properly with fetched data 1`] = `
<payments-list-core>
  <div>
    <payments-list-filters></payments-list-filters>
    <div class="table-wrapper">
      <table class="table table-hover">
        <thead class="sticky-top table-head" part="table-head">
          <tr class="table-light text-nowrap" part="table-head-row">
            <th part="table-head-cell" scope="col" title="The date and time each payment was made">
              Date
            </th>
            <th part="table-head-cell" scope="col" title="The dollar amount of each payment">
              Amount
            </th>
            <th part="table-head-cell" scope="col" title="The current status of each payment">
              Status
            </th>
            <th part="table-head-cell" scope="col" title="The type of each payment">
              Type
            </th>
            <th part="table-head-cell" scope="col" title="The payment description, if you provided one">
              Description
            </th>
            <th part="table-head-cell" scope="col" title="The name associated with the payment method">
              Account Holder
            </th>
            <th part="table-head-cell" scope="col" title="The brand and last 4 digits of the payment method">
              Payment Method
            </th>
          </tr>
        </thead>
        <tbody class="table-body" part="table-body">
          <tr data-row-entity-id="py_1A2B3C4D5E6F7G8H9I0J" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                May 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $123.45 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-success" title="This payment was successfully captured">
                Succeeded
              </span>
            </td>
            <td part="table-cell">
              Card
            </td>
            <td part="table-cell">
              one-time payment
            </td>
            <td part="table-cell">
              test card
            </td>
            <td part="table-cell">
              **** 1234
            </td>
          </tr>
          <tr data-row-entity-id="py_2B3C4D5E6F7G8H9I0J1K" data-test-id="table-row" part="table-row table-row-even">
            <td part="table-cell">
              <div class="fw-bold">
                Jun 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $234.56 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-primary" title="This ACH payment was processed, but the funds have not settled. It could still succeed or fail.">
                Pending
              </span>
            </td>
            <td part="table-cell">
              ACH
            </td>
            <td part="table-cell">
              subscription payment
            </td>
            <td part="table-cell">
              test bank account
            </td>
            <td part="table-cell">
              **** 5678
            </td>
          </tr>
          <tr data-row-entity-id="py_3C4D5E6F7G8H9I0J1K2L" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Jul 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $345.67 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-success" title="This payment was successfully captured">
                Succeeded
              </span>
            </td>
            <td part="table-cell">
              Card
            </td>
            <td part="table-cell">
              manual payment
            </td>
            <td part="table-cell">
              test card
            </td>
            <td part="table-cell">
              **** 9101
            </td>
          </tr>
          <tr data-row-entity-id="py_1A2B3C4D5E6F7G8H9I0J" data-test-id="table-row" part="table-row table-row-even">
            <td part="table-cell">
              <div class="fw-bold">
                May 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $123.45 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-danger" title="This card payment did not go through (it will not appear in a payout)">
                Failed
              </span>
            </td>
            <td part="table-cell">
              Card
            </td>
            <td part="table-cell">
              one-time payment
            </td>
            <td part="table-cell">
              test card
            </td>
            <td part="table-cell">
              **** 1234
            </td>
          </tr>
          <tr data-row-entity-id="py_2B3C4D5E6F7G8H9I0J1K" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Jun 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $234.56 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-primary" title="This ACH payment was processed, but the funds have not settled. It could still succeed or fail.">
                Pending
              </span>
            </td>
            <td part="table-cell">
              ACH
            </td>
            <td part="table-cell">
              subscription payment
            </td>
            <td part="table-cell">
              test bank account
            </td>
            <td part="table-cell">
              **** 5678
            </td>
          </tr>
          <tr data-row-entity-id="py_3C4D5E6F7G8H9I0J1K2L" data-test-id="table-row" part="table-row table-row-even">
            <td part="table-cell">
              <div class="fw-bold">
                Jul 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $345.67 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-secondary" title="The account holder disputed this payment. The amount has been returned and a fee assessed.">
                Disputed
              </span>
            </td>
            <td part="table-cell">
              Card
            </td>
            <td part="table-cell">
              manual payment
            </td>
            <td part="table-cell">
              test card
            </td>
            <td part="table-cell">
              **** 9101
            </td>
          </tr>
          <tr data-row-entity-id="py_4D5E6F7G8H9I0J1K2L3M" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 1, 2024
              </div>
              <div class="fw-bold">
                8:45am
              </div>
            </td>
            <td part="table-cell">
              $345.67 USD
            </td>
            <td part="table-cell">
              <span class="badge bg-secondary" title="The full amount of this payment has been refunded">
                Fully Refunded
              </span>
            </td>
            <td part="table-cell">
              Card
            </td>
            <td part="table-cell">
              refund payment
            </td>
            <td part="table-cell">
              test card
            </td>
            <td part="table-cell">
              **** 1122
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="7" part="pagination-bar">
              <pagination-menu>
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination" part="pagination">
                    <li class="disabled page-item" part="page-item">
                      <a class="page-link" href="#" part="page-link page-link-disabled">
                        <span class="me-1" part="page-link-arrow">
                          «
                        </span>
                        <span part="page-button-text">
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-link page-link-disabled">
                        <span part="page-link-text">
                          Next
                        </span>
                        <span class="ms-1" part="page-link-arrow">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</payments-list-core>
`;
